components/builder-db/src/error.rs

```diff
@@ -41,7 +41,6 @@ pub enum Error {
     SchemaDrop(postgres::error::Error),
     SchemaSwitch(postgres::error::Error),
     SetSearchPath(postgres::error::Error),
-    ShardMigrationIncomplete,
```

Comment:
I'm not sure we can remove this since we have potential on-prem instances that might have un-migrated data.

Reply:
If we leave this we will have db table collision between the services how big of an upgrade do we need to support? Maybe we could create a touch file instead?

Reply:
Where is the db table collision happening?  I don't see the flags table in the merge script..

Reply:
That's because I don't migrate the flags table to avoid the collision

Reply:
As discussed if the migration script checks for the flags table and requires a migration before proceeding then we should be fine not migrating the flags table.  But we still must leave this check in.  If an on-prem user uninstalls/reinstalls, and gets the latest services, the services will start normally if the check is not there.  That will mess up their env.  So the shard migration check needs to be kept in the code

Reply:
The way the tables are designed it's not possible to leave them in safely. They all have the same table name and the same column name to check if a migration is complete. If you migrate one of the flags tables and leave this code in it doesn't mean that the other two databases have their shards migrated. What if we write a script that creates a touch file in the services mutable directory based on the table in the db? Then we can put part of this code back and rely on the touch-file to see if the migration is complete

Reply:
I think the touch file approach is flawed in key ways - we add state to the services directory that needs to be carried around (I'm against doing this - thats what the datastore is for), and we create yet another way to give the services an "all clear" to start.  The scenario you are concerned about (a partial shard migration) is not a real world concern - let's say an on-prem admin runs shard-migrate.sh and it fails (partial migration).  Well, they are not going to be able to run your merge-databases.sh script (since you're going to check for proper flags for all services).  So at that point, what an admin will need to do is 1) run shard migration, then 2) run db migration.  Then, services will come up properly.  These 2 steps are exactly what we will document in the on-prem instructions.

Reply:
The one thing that is tricky is that we added the check for no rows to specifically let brand new DBs start without the shard error.  So we have to think about how that is going to play out as well.  If we keep that as is, then there's no way to stop the on-prem install from bringing up new services and immediately starting with a fresh DB after this goes in :(

Reply:
I can't think of any great solutions to this.  We may have to just put in a visible warning in the on-prem readme to shut down services immediately after re-install and do the db merge.  The db merge script will then have to check for the scenario of non-empty DB and and not assume that merge  db was run - instead it should halt the merge.